src/v/archival/archival_policy.cc

```diff
@@ -287,6 +234,136 @@ static ss::output_stream<char> make_null_output_stream() {
+    model::timestamp ts = base_timestamp;
+    auto res = co_await storage::transform_stream(
+      reader_handle.take_stream(),
+      std::move(ostr),
+      [begin_inclusive, &sto, &ts](model::record_batch_header& hdr) {
+          if (hdr.last_offset() < begin_inclusive) {
+              // The current record batch is accepted and will contribute to
+              // skipped length. This means that if we will read segment
+              // file starting from the 'scan_from' + 'res' we will be
+              // looking at the next record batch. We might not see the
+              // offset that we're looking for in this segment. This is why
+              // we need to update 'sto' per batch.
+              sto = hdr.last_offset() + model::offset(1);
+              // TODO: update base_timestamp
+              return storage::batch_consumer::consume_result::accept_batch;
+          }
+          ts = hdr.first_timestamp;
+          return storage::batch_consumer::consume_result::stop_parser;
+      });
+    co_await reader_handle.close();
+    if (res.has_error()) {
+        vlog(
+          archival_log.error,
+          "Can't read segment file, error: {}",
+          res.error());
+        throw std::system_error(res.error());
+    } else {
+        bytes_to_skip = scan_from + res.value();
+        vlog(
+          archival_log.debug,
+          "Scanned {} bytes starting from {}, total {}. Adjusted starting "
+          "offset: {}",
+          res.value(),
+          scan_from,
+          bytes_to_skip,
+          sto);
+    }
+    // Adjust content length and offsets at the begining of the file
+    co_return offset_to_file_pos_result{sto, bytes_to_skip, ts};
+}
+ss::future<offset_to_file_pos_result> convert_end_offset_to_file_pos(
+  model::offset end_inclusive,
+  const ss::lw_shared_ptr<storage::segment>& segment,
+  model::timestamp max_timestamp,
+  ss::io_priority_class io_priority) {
+    // Handle truncated segment upload (if the upload was triggered by time
+    // limit). Note that the upload is not necessarily started at the beginning
+    // of the segment.
+    // Lookup the index, if the index is available and some value is found
+    // use it as a starting point otherwise, start from the beginning.
+    auto ix_end = segment->index().find_nearest(end_inclusive);
+    size_t fsize = segment->reader().file_size();
+
+    // NOTE: Index lookup might return an offset which isn't committed yet.
+    // Subsequent call to segment_reader::data_stream will fail in this
+    // case. In order to avoid this we need to make another index lookup
+    // to find a lower offset which is committed.
+    while (ix_end && ix_end->filepos >= fsize) {
+        vlog(archival_log.debug, "The position is not flushed {}", *ix_end);
+        auto lookup_offset = ix_end->offset - model::offset(1);
+        ix_end = segment->index().find_nearest(lookup_offset);
+        vlog(archival_log.debug, "Re-adjusted position {}", *ix_end);
```

Comment:
The first

>        vlog(archival_log.debug, "The position is not flushed {}", *ix_end);

has a check on the dereference performed by the `while-loop` predicate. but the second one is an unqualified dereference.

```
        ix_end = segment->index().find_nearest(lookup_offset);
        vlog(archival_log.debug, "Re-adjusted position {}", *ix_end);
```

is this safe? does `find_nearest` always return an iterator that is valid to dereference?

Reply:
Thanks @abhijat !